content/references/rippled-api/api-conventions/serialization.md

```diff
@@ -0,0 +1,285 @@
+
+The [canonical field order](#canonical-field-order) of object fields is the same as the canonical field order for all top-level fields, but the members of the object must be sorted within the object. After the last member, there is an "Object end" Field ID of `0xe1` with no contents.
+
+The following example shows the serialization format for an object (a single `Memo` object in the `Memos` array).
+
+![Object field ID, followed by the Object ID and contents of each object member in canonical order, followed by the "Object end" field ID](img/serialization-object.png)
+
+
+### PathSet Fields
+[PathSet]: #pathset-fields
+
+The `Paths` field of a cross-currency [Payment transaction][] is a "PathSet", represented in JSON as an array of arrays. For more information on what paths are used for, see [Paths](paths.html).
+
+A PathSet is serialized as **1 to 6** individual paths in sequence[[Source]](https://github.com/ripple/rippled/blob/4cff94f7a4a05302bdf1a248515379da99c5bcd4/src/ripple/app/tx/impl/Payment.h#L35-L36 "Source"). Each complete path is followed by a byte that indicates what comes next:
+
+- `0xff` indicates another path follows
+- `0x00` indicates the end of the PathSet
+
+Each path consists of **1 to 8** path steps in order[[Source]](https://github.com/ripple/rippled/blob/4cff94f7a4a05302bdf1a248515379da99c5bcd4/src/ripple/app/tx/impl/Payment.h#L38-L39 "Source"). Each step starts with a **type** byte, followed by one or more fields describing the path step. The type indicates which fields are present in that path step through bitwise flags. (For example, the value `0x30` indicates changing both currency and issuer.) If more than one field is present, the fields are always placed in a specific order.
+
+The following table describes the possible fields and the bitwise flags to set in the type byte to indicate them:
+
+| Type Flag | Field Present | Field Type        | Bit Size | Order |
+|:----------|:--------------|:------------------|:---------|:------|
+| `0x01`    | `account`     | [AccountID][]     | 160 bits | 1st   |
+| `0x10`    | `currency`    | [Currency Code][] | 160 bits | 2nd   |
+| `0x20`    | `issuer`      | [AccountID][]     | 160 bits | 3rd   |
+
+[Currency Code]: currency-formats.html#standard-currency-codes
+
+Some combinations are invalid; see [Path Specifications](paths.html#path-specifications) for details.
+
+The AccountIDs in the `account` and `issuer` fields are presented _without_ a variable-length encoding prefix. When the `currency` is XRP, the currency code is represented as 160 bits of zeroes.
+
+Each step is followed directly by the next step of the path. As described above, last step of a path is followed by either `0xff` (if another path follows) or `0x00` (if this ends the last path).
+
+The following example shows the serialization format for a PathSet:
+
```

Comment:
In this graphic, curious about how the Type flag values for Currency (160 bits) and Issuer (160 bits) are "combined" to use a single Type flag of 0x30. I am in way over my head here, but I wasn't thinking that Type flag values could be "added" - as in 0x10 + 0x20 = 0x30 -- is that what is happening here?

Reply:
The type object has a bunch of flags that are combined with bitwise OR operations.

```
0x20 in binary:
00100000

0x10 in binary:
00010000

    00100000
OR  00010000
------------
    00110000  = 0x30 in hex

Reply:
Thank you for taking the time to explain and illustrate...this is 🏅!